database/src/main/postgres/runs/V1.5.12__create_or_update_additional_data.sql

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright 2021 ABSA Group Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+CREATE OR REPLACE FUNCTION runs.create_or_update_additional_data(
+    IN  i_partitioning      JSONB,
+    IN  i_additional_data   HSTORE,
+    IN  i_by_user           TEXT,
+    OUT status              INTEGER,
+    OUT status_text         TEXT,
+    OUT id_additional_data  BIGINT
+) RETURNS record AS
+$$
+-------------------------------------------------------------------------------
+--
+-- Function: runs.create_or_update_additional_data(3)
+--      Adds the additional data for the input partitioning. If additional data of a given name already
+--      exists for such partitioning, the value is updated and the old value is moved to the
+--      additional data history table.
+--
+-- Parameters:
+--      i_partitioning      - partitioning to add the additional data for
+--      i_additional_data   - sets of key/value pairs representing name and values of the additional data
+--      i_by_user           - user behind the change
+--
+-- Returns:
+--      status              - Status code
+--      status_text         - Status text
+--      id_additional_data  - id of the data added
+--
+-- Status codes:
+--      11                  - Additional data have been added
+--      12                  - Additional data have been upserted
+--      14                  - No changes in additional data (this is when they already existed)
+--      41                  - Partitioning not found
+--
+-------------------------------------------------------------------------------
+DECLARE
+    _fk_partitioning        BIGINT;
+    _ad_backup_performed    BOOLEAN;
+BEGIN
+
+    _fk_partitioning := runs._get_id_partitioning(i_partitioning, true);
+
+    IF _fk_partitioning IS NULL THEN
+        status := 41;
+        status_text := 'Partitioning not found';
+        RETURN;
+    END IF;
+
+    -- 1. (backup) get records that already exist but values differ, and insert them into ad history table
+    INSERT INTO runs.additional_data_history
```

Comment:
Is there even a reason to do a backup in this sort of way? We can always return only the latest value for a given key in a select query for the additional data.

Reply:
It's definitely technically possible and I was even considering it. But I disregarded such idea, because 
* over time such (one) table would include only a small fraction of data that would be actual, and most would be historical - the queries would always need to work with such information - so the queries and indexes would be a bit more complicated
* there are different data constraints on historical data and actual data - I'm more strict to actual data - such as, I never want to have multiple value&key assigned to a given partitioning; but in history it's not only tolerated but it's also encouraged - e.g. an AD record (A, B)  can be created, then updated to (A, C), and then updated back to (A, B) and then updated to (A, D) - so the history table will contain (A, B) pair twice (with different metadata ofc). But I never wanted to have (A, B) twice in the main table (again, given exactly the same partitioning)

Reply:
I see, okey sounds reasonable then, thanks for explanation.